client/X11/xf_rail.c

```diff
@@ -531,22 +548,337 @@ static BOOL xf_rail_window_delete(rdpContext* context,
+			if (!nextBit)
+			{
+				nextBit = 0x80;
+				maskByte++;
+			}
+		}
+	}
+}
+
+/*
+ * _NET_WM_ICON format is defined as "array of CARDINAL" values which for
+ * Xlib must be represented with an array of C's "long" values. Note that
+ * "long" != "INT32" on 64-bit systems. Therefore we can't simply cast
+ * the bitmap data as (unsigned char*), we have to copy all the pixels.
+ *
+ * The first two values are width and height followed by actual color data
+ * in ARGB format (e.g., 0xFFFF0000L is opaque red), pixels are in normal,
+ * left-to-right top-down order.
+ */
+static BOOL convert_rail_icon(ICON_INFO* iconInfo, xfRailIcon *railIcon)
+{
+	BYTE* argbPixels;
+	BYTE* nextPixel;
+	long* pixels;
+	int i;
+	int nelements;
+
+	argbPixels = calloc(iconInfo->width * iconInfo->height, 4);
+	if (!argbPixels)
+		goto error;
+
+	if (!convert_icon_color_to_argb(iconInfo, argbPixels))
+		goto error;
+
+	apply_icon_alpha_mask(iconInfo, argbPixels);
+
+	nelements = 2 + iconInfo->width * iconInfo->height;
+	pixels = realloc(railIcon->data, nelements * sizeof(*pixels));
+	if (!pixels)
+		goto error;
+
+	railIcon->data = pixels;
+	railIcon->length = nelements;
+
+	pixels[0] = iconInfo->width;
+	pixels[1] = iconInfo->height;
+
+	nextPixel = argbPixels;
+	for (i = 2; i < nelements; i++)
+	{
+		pixels[i] = read_color_quad(nextPixel);
+		nextPixel += 4;
+	}
+
+	free(argbPixels);
+
+	return TRUE;
+
+error:
+	free(argbPixels);
+	return FALSE;
+}
+
+static void xf_rail_set_window_icon(xfContext* xfc,
+                                    xfAppWindow* railWindow, xfRailIcon *icon,
+                                    BOOL replace)
+{
+	XChangeProperty(xfc->display, railWindow->handle, xfc->_NET_WM_ICON,
+		XA_CARDINAL, 32, replace ? PropModeReplace : PropModeAppend,
+		(unsigned char*) icon->data, icon->length);
+	XFlush(xfc->display);
+}
+
+static xfAppWindow* xf_rail_get_window_by_id(xfContext* xfc, UINT32 windowId)
+{
+	return (xfAppWindow*) HashTable_GetItemValue(xfc->railWindows,
+		(void*)(UINT_PTR) windowId);
+}
+
 static BOOL xf_rail_window_icon(rdpContext* context,
                                 WINDOW_ORDER_INFO* orderInfo, WINDOW_ICON_ORDER* windowIcon)
 {
 	xfContext* xfc = (xfContext*) context;
-	xfAppWindow* railWindow = (xfAppWindow*) HashTable_GetItemValue(xfc->railWindows,
-	                          (void*)(UINT_PTR) orderInfo->windowId);
+	xfAppWindow* railWindow;
+	xfRailIcon *icon;
+	BOOL replaceIcon;
 
+	railWindow = xf_rail_get_window_by_id(xfc, orderInfo->windowId);
 	if (!railWindow)
+		return TRUE;
+
+	icon = RailIconCache_Lookup(xfc->railIconCache,
+		windowIcon->iconInfo->cacheId,
+		windowIcon->iconInfo->cacheEntry);
+	if (!icon)
+	{
+		WLog_DBG(TAG, "failed to get icon from cache %02X:%04X",
+			windowIcon->iconInfo->cacheId,
+			windowIcon->iconInfo->cacheEntry);
 		return FALSE;
+	}
+
+	if (!convert_rail_icon(windowIcon->iconInfo, icon))
+	{
+		WLog_DBG(TAG, "failed to convert icon for window %08X", orderInfo->windowId);
```

Comment:
> why not abort here if the data was invalid?

Mmm... don't remember. I think I wanted the malformed icon to be set to the window so that I see that the conversion has been made incorrectly. However, just hiding malformed icons is a resonable choice too, so returning from the function should be okay.